types/react-redux/index.d.ts

```diff
@@ -1,35 +1,50 @@
 // Type definitions for react-redux 4.4.47
 // Project: https://github.com/rackt/react-redux
-// Definitions by: Qubo <https://github.com/tkqubo>, Sean Kelley <https://github.com/seansfkelley>, Thomas Hasner <https://github.com/thasner>
+// Definitions by: Qubo <https://github.com/tkqubo>,
+//                 Sean Kelley <https://github.com/seansfkelley>,
+//                 Thomas Hasner <https://github.com/thasner>
+//                 Curits Layne <https://github.com/clayne11>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+// TypeScript Version: 2.4
 
 import * as React from 'react';
 import * as Redux from 'redux';
 
 type ComponentClass<P> = React.ComponentClass<P>;
 type StatelessComponent<P> = React.StatelessComponent<P>;
-type Component<P> = ComponentClass<P> | StatelessComponent<P>;
+type Component<P> = React.ComponentType<P>;
 type ReactNode = React.ReactNode;
 type Store<S> = Redux.Store<S>;
 type Dispatch<S> = Redux.Dispatch<S>;
 type ActionCreator<A> = Redux.ActionCreator<A>;
 
+// Diff / Omit taken from https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-311923766
+type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
+type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
+
 export interface DispatchProp<S> {
   dispatch: Dispatch<S>;
 }
 
-interface ComponentDecorator<TMergedProps, TOwnProps> {
-    (component: Component<TOwnProps & TMergedProps>): ComponentClass<TOwnProps>;
+// Injects props and removes them from the prop requirements.
+// Will not pass through the injected props if they are passed in during
+// render. Also adds new prop requirements from TNeedsProps.
+export interface InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> {
+    <P extends TInjectedProps>(
+        component: Component<P>
+    ): ComponentClass<Omit<P, keyof TInjectedProps> & TNeedsProps>
 }
 
-interface ComponentDecoratorInfer<TMergedProps> {
-    <T>(component: Component<T & TMergedProps>): ComponentClass<T>;
-}
+// Injects props and removes them from the prop requirements.
+// Will not pass through the injected props if they are passed in during
+// render.
+export type InferableComponentEnhancer<TInjectedProps> =
+    InferableComponentEnhancerWithProps<TInjectedProps, {}>
 
-interface ComponentMergeDecorator<TMergedProps, TOwnProps> {
-    (component: Component<TMergedProps>): ComponentClass<TOwnProps>;
-}
+// Injects default props and makes them optional. Will still pass through
+// the injected props if they are passed in during render.
+export type DefaultingInferableComponentEnhancer<TInjectedProps> =
```

Comment:
Nope good catch - I accidentally brought that over from other types I was working on. I'll remove it.